src/java/org/apache/cassandra/cql3/Terms.java

```diff
@@ -47,225 +47,510 @@ public Object get(int index)
+         * specification and "prepare" this {@code Terms.Raw}, returning the resulting
+         *
+         * @param receiver the "column" the set of terms are supposed to be a value of. Note
+         * that the ColumnSpecification may not correspond to a real column.
+         * @return the prepared terms
+         */
+        public abstract Terms prepare(String keyspace, ColumnSpecification receiver) throws InvalidRequestException;
+
+        /**
+         * @return a String representation of the raw terms that can be used when reconstructing a CQL query string.
+         */
+        public abstract String getText();
+
+        /**
+         * The type of the {@code Terms} if it can be infered.
+         *
+         * @param keyspace the keyspace on which the statement containing these terms is on.
+         * @return the type of this {@code Terms} if inferrable, or {@code null}
+         * otherwise (for instance, the type isn't inferrable for a bind marker. Even for
+         * literals, the exact type is not inferrable since they are valid for many
+         * different types and so this will return {@code null} too).
+         */
+        public abstract AbstractType<?> getExactTypeIfKnown(String keyspace);
+
+        @Override
+        public AbstractType<?> getCompatibleTypeIfKnown(String keyspace)
+        {
+            return getExactTypeIfKnown(keyspace);
+        }
+
+        @Override
+        public String toString()
+        {
+            return getText();
+        }
+
+        public static Raw of(List<? extends Term.Raw> raws)
+        {
+            return new Raw()
             {
-                Terminal terminal = marker.bind(options);
+                @Override
+                public Terms prepare(String keyspace, ColumnSpecification receiver) throws InvalidRequestException
+                {
+                    List<Term> terms = new ArrayList<>(raws.size());
+                    for (Term.Raw raw : raws)
+                    {
+                        terms.add(raw.prepare(keyspace, receiver));
+                    }
+                    return Terms.of(terms);
+                }
+
+                @Override
+                public String getText()
+                {
+                    return raws.stream().map(Term.Raw::getText)
+                                        .collect(Collectors.joining(", ", "(", ")"));
+                }
 
-                if (terminal == null)
+                @Override
+                public AbstractType<?> getExactTypeIfKnown(String keyspace)
+                {
                     return null;
+                }
+
+                @Override
+                public TestResult testAssignment(String keyspace, ColumnSpecification receiver)
+                {
+                    return AssignmentTestable.TestResult.WEAKLY_ASSIGNABLE;
+                }
+            };
+        }
+    }
+
+    abstract class Terminals implements Terms
```

Comment:
It might be useful to clarify in the class JavaDoc that all the terms in this class are expected to be terminals.